src/main/java/gregtech/api/recipes/RecipeMap.java

```diff
@@ -327,95 +296,346 @@ public Recipe findRecipe(long voltage, List<ItemStack> inputs, List<FluidStack>
-                        if (!exactVoltage && voltage < tmpRecipe.getEUt()) {
-                            continue;
-                        } else if (exactVoltage && voltage != tmpRecipe.getEUt()) {
-                            continue;
-                        }
-                        calculateRecipePriority(tmpRecipe, promotedTimes, priorityRecipeMap);
-                    }
+    /**
+     * Builds a list of unique ItemStacks from the given Collection of ItemStacks.
+     * Used to reduce the number inputs, if for example there is more than one of the same input,
+     * pack them into one.
+     * This uses a strict comparison, so it will not pack the same item with different NBT tags,
+     * to allow the presence of, for example, more than one configured circuit in the input.
+     * @param input The Collection of GTRecipeInputs.
+     * @return an array of unique itemstacks.
+     */
+
+    public static ItemStack[] uniqueItems(Collection<ItemStack> input) {
+        List<ItemStack> list = new ObjectArrayList<>(input.size());
+        loop:
+        for (ItemStack item : input) {
+            if (item.isEmpty()) {
+                continue;
+            }
+            for (ItemStack obj : list) {
+                if (item.isItemEqual(obj) && ItemStack.areItemStackTagsEqual(item, obj)) {
+                    continue loop;
                 }
             }
+            list.add(item);
         }
+        return list.toArray(new ItemStack[0]);
+    }
 
-        for (FluidStack fluidStack : fluidInputs) {
-            if (fluidStack != null) {
-                FluidKey fluidKey = new FluidKey(fluidStack);
-                if (!searchedFluids.contains(fluidKey) && recipeFluidMap.containsKey(fluidKey)) {
-                    searchedFluids.add(fluidKey);
-                    for (Recipe tmpRecipe : recipeFluidMap.get(fluidKey)) {
-                        if (!exactVoltage && voltage < tmpRecipe.getEUt()) {
-                            continue;
-                        } else if (exactVoltage && voltage != tmpRecipe.getEUt()) {
-                            continue;
-                        }
-                        calculateRecipePriority(tmpRecipe, promotedTimes, priorityRecipeMap);
-                    }
+    /**
+     * Builds a list of unique inputs from the given list GTRecipeInputs.
+     * Used to reduce the number inputs, if for example there is more than one of the same input,
+     * pack them into one.
+     * @param input The list of GTRecipeInputs.
+     * @return The list of unique inputs.
+     */
+
+    public static List<GTRecipeInput> uniqueIngredientsList(List<GTRecipeInput> input) {
+        List<GTRecipeInput> list = new ObjectArrayList<>(input.size());
+        loop:
+        for (GTRecipeInput item : input) {
+            for (GTRecipeInput obj : list) {
+                if (item.equalIgnoreAmount(obj)) {
+                    continue loop;
                 }
             }
+            list.add(item);
         }
```

Comment:
You can de-label this segment with the following:

```suggestion
        for (GTRecipeInput item : input) {
            boolean isEqual = false;
            for (GTRecipeInput obj : list) {
                if (item.equalIgnoreAmount(obj)) {
                    isEqual = true;
                    break;
                }
            }
            if (isEqual) continue;
            list.add(item);
        }
```